Add in documentation on APIv4 and Custom Data (#695)
civicrm/civicrm-dev-docs: docs/api/v4/custom-data.md

# APIv4 and Custom Data

CiviCRM has a flexible custom data system which allows nearly any entity to be extended. It comes in two distinct flavors: Single-record and Multi-record. For more background see the user guide chapter: [Creating Custom Fields](https://docs.civicrm.org/user/en/latest/organising-your-data/creating-custom-fields/).

> ## Single-record custom data

Reviewer: ```suggestion
## Single-Record Custom Data
```

> Because single-record fields extend an entity 1-to-1, the api treats the custom fields as an extension of the regular fields. For example, normally an Event has fields like `id`, `title`, `start_date`, `end_date`, etc. Adding a custom group named "Event_Extra" and a field named "Room_number" would be accessable from the api as "Event_Extra.Room_number". You would retrieve it and set it as if it were any other field. Note that the `name` of a field is not to be confused with the `label` The api refers to custom groups/fields by name and not user-facing labels which are subject to translation and alteration.

Reviewer: ```suggestion
Because single-record fields extend an entity 1-to-1, the API treats the custom fields as an extension of the regular fields. For example, normally an Event has fields like `id`, `title`, `start_date`, `end_date`, etc. Adding a custom group named "Event_Extra" and a field named "Room_number" would be accessible from the API as "Event_Extra.Room_number". You would retrieve it and set it as if it were any other field. Note that the `name` of a field is not to be confused with the `label` The API refers to custom groups/fields by name and not user-facing labels which are subject to translation and alteration.
```

For setting custom date fields, date format is anything understood by `strtotime`, e.g. "now" or "-1 week" or "2020-12-25".

> ## Multi-record custom data

Reviewer: ```suggestion
## Multi-Record Custom Data
```

> Multiple record custom data sets are treated by the api as entities, which work similarly to other entities attached to contacts (Phone, Email, Address, etc.). For example, creating a multi-record set named "Work_history" could then be accessed via the api as an entity named "Custom_Work_history". Creating a record would be done like so:

Reviewer: ```suggestion
Multiple record custom data sets are treated by the API as entities, which work similarly to other entities attached to contacts (Phone, Email, Address, etc.). For example, creating a multi-record set named "Work_History" could then be accessed via the API as an entity named "Custom_Work_History". Creating a record would be done like so:
```

```php
civicrm_api4('Custom_Work_history', 'create', $params);
```

> ## Try it out

Reviewer: ```suggestion
## Try It Out
```

> Once you have created some custom data in your system, look for it in the api explorer. Single-record data will appear as fields on the entities they extend, and multi-record data will appear in the list of entities (look under "C" alphabetically as they all start with the prefix "Custom_".

Reviewer: ```suggestion
Once you have created some custom data in your system, look for it in the API explorer. Single-record data will appear as fields on the entities they extend, and multi-record data will appear in the list of entities (look under "C" alphabetically as they all start with the prefix "Custom_".
```

# APIv4 and Custom Data

Custom data attached to entities is referenced in the format of `custom_group_machine_name.custom_field_machine_name`.

To set a custom field, or find entities with custom fields of a particular value, you typically use a parameter like this:

```php
$values['custom_field_reference'] = 'value';
```

To return custom data for an entity just include the machine name of the custom data in the select array.

> For setting custom date fields, (ie CustomValue create), date format is `YmdHis`, for example: `20050425000000`.

Reviewer 2: ```suggestion
For setting custom date fields, date format is anything understood by `strtotime`, e.g. "now" or "-1 week" or "2020-12-25"
```

> This is just a brief introduction; each API may have different requirements and allow different formats for accessing the custom data. See the [API function documentation](/api/index.md) and also read the comments and documentation in each API php file (under civicrm/CRM/api/v3 in your CiviCRM installation) for exact details,

Reviewer: ```suggestion
This is just a brief introduction; each API may have different requirements and allow different formats for accessing the custom data. See the [API function documentation](/api/index.md) and also read the comments and documentation in each API php file (under `civicrm/CRM/api/v3` in your CiviCRM installation) for exact details,
```

which vary for each API entity and function.

!!! note 
    When retrieving custom data from a multiple record custom group set the custom data will be returned in an array of custom field which contains the value and the id of the row in the custom field table.

> !!! note

Author: @colemanw if you are able to enlighten me here that would be appreciated

Reviewer 2: Multivalue custom data is treated as an entity in api4, so you would set multivalue data using e.g. `civicrm_api4('Custom_My_Cool_Fields', 'create, $params)`
